### PR TITLE
QLoRA Support + NNX Decoder Sharding Fixes

### PR DESCRIPTION
**Pull request by RexBearIU**
# Description

This PR introduces core support for QLoRA and implements robust sharding metadata synchronization for NNX decoders.

  Currently, applying LoRA adapters alongside quantization and complex multi-host sharding setups can lead to PartitionSpec mismatches and cross-backend device_put issues during lax.scan. 

  This PR solves these issues by:
   - Extending the LoRA configuration to support quantization types (lora_weight_qtype) and block sizes (lora_tile_size) for QLoRA.
   - Adding generic NNX metadata synchronization helpers (nnx_remove_scan_axis, nnx_add_scan_axis, nnx_sync_moveaxis) to safely manipulate PartitionSpec metadata during lax.scan operations in NNXDecoder.
   - Implementing a _safe_reshard workaround using jax.make_array_from_callback in lora_utils.py to natively construct globally sharded arrays, bypassing backend-specific device_put issues on Pathways/McJAX.

  Future improvements will include removing the _safe_reshard workaround once the underlying JAX/Qwix parameter materialization issues are fully resolved upstream.

# Tests

 - Unit tests added for NNX metadata synchronization (tests/utils/test_maxtext_utils_nnx.py).
   - Existing unit tests for lora_utils updated (note: QLoRA specific tests are temporarily skipped pending an upstream qwix fix).
   - Tested via multi-host mock execution to trigger the resharding callback logic.

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [X] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [X] I have necessary comments in my code, particularly in hard-to-understand areas.
- [X] I have run end-to-end tests tests and provided workload links above if applicable.
- [X] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/maxtext/configs/types.py [hunk 1]
```diff
@@ -1740,6 +1750,14 @@ class HloDump(BaseModel):
   dump_jaxpr_gcs_dir: PathStr = Field("", description="GCS directory to upload jaxpr dumps.")
 
 
+class StackTrace(BaseModel):
```
Comment (SurbhiJainUSC) [inline src/maxtext/configs/types.py]: This has been removed from main. Please rebase.

File: src/maxtext/configs/types.py [hunk 2]
```diff
@@ -2313,6 +2331,7 @@ class MaxTextConfig(
     DevelopmentAndDebugging,
     Profiling,
     HloDump,
+    StackTrace,
```
Comment (SurbhiJainUSC) [inline src/maxtext/configs/types.py]: Remove this

File: tests/post_training/unit/lora_utils_test.py [hunk 3]
```diff
@@ -200,13 +215,27 @@ def _run_apply_lora_test(self, scan_layers: bool):
...
-    """Test applying LoRA to model with scan_layers=False."""
+    """Test applying standard LoRA to model with scan_layers=False."""
     self._run_apply_lora_test(scan_layers=False)
 
   def test_apply_lora_to_model_scan_layers_true(self):
-    """Test applying LoRA to model with scan_layers=True."""
+    """Test applying standard LoRA to model with scan_layers=True."""
     self._run_apply_lora_test(scan_layers=True)
 
+  @unittest.skip("Awaiting qwix fix for QLoRA params materialization")
```
Comment (SurbhiJainUSC) [inline tests/post_training/unit/lora_utils_test.py]: Why are we skipping QLoRA tests? Does that mean QLoRA won't work even after merging this PR?

File: src/maxtext/configs/types.py (hunk 1 above)
Comment (RexBearIU) [reply src/maxtext/configs/types.py]: Done. I have rebased the branch onto the latest  main  and completely removed the  StackTrace  configurations.

File: src/maxtext/configs/types.py (hunk 2 above)
Comment (RexBearIU) [reply src/maxtext/configs/types.py]: Done! It has been removed.

File: tests/post_training/unit/lora_utils_test.py (hunk 3 above)
Comment (RexBearIU) [reply tests/post_training/unit/lora_utils_test.py]: Yes, we need a new qwix release for NNX model QLoRA support. 
  
My two CLs addressing the underlying parameter materialization issues have already been merged upstream in qwix. I will sync with the team to ask for their next release plan. Once the new release comes out, we just need to update the dependency version here and unskip these tests.
  
Should we wait and update the dependency/unskip tests in this PR, or would you prefer to merge this PR as-is and do the update in a separate follow-up PR once the release is published?

Comment (SurbhiJainUSC) [reply tests/post_training/unit/lora_utils_test.py]: I would say let's wait for Qwix fix, update dependency in MAxText and then merge this PR.

Comment (RexBearIU) [reply tests/post_training/unit/lora_utils_test.py]: Will do. I'll leave this PR open and update the qwix dependency as soon as the new release is published.

**Review by shralex — APPROVED** [review]

File: src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt [hunk 4]
```diff
@@ -140,9 +140,9 @@ ipython-pygments-lexers>=1.1.1
 ipywidgets>=8.1.8
 isort>=8.0.1
 jaraco-functools>=4.5.0
-jax>=0.10.1
+jax>=0.10.2
```
Comment (SurbhiJainUSC) [inline src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt]: Should jax/jaxlib have the same version?

File: src/maxtext/utils/maxtext_utils_nnx.py [hunk 5]
```diff
@@ -187,3 +188,90 @@ def _op(x):
...
+    is_var = isinstance(x, nnx.Variable)
+    val = x.get_value() if is_var else x
+    if not hasattr(val, "shape"):
+      return x
+
+    new_val = jnp.moveaxis(val, from_axis, to_axis)
+    if not is_var:
+      return new_val
+
+    def move_fn(l):
```
Comment (khatwanimohit) [inline src/maxtext/utils/maxtext_utils_nnx.py:232]: If a model variable's sharding specification omits trailing `None`s (which is very common, e.g. `PartitionSpec("layers", "embed")` for a 3D tensor of shape `(10, 20, 30)`), the condition `len(l) >
  max(from_axis, to_axis)` evaluates to `False`.

We should pad the PartitionSpec list `l` to the full rank of the array (`val.ndim`) by appending `None` to the end *before* moving the axis.

```suggestion
    def move_fn(l):
      while len(l) < val.ndim:
        l.append(None)
      if len(l) > max(from_axis, to_axis):
        l.insert(to_axis, l.pop(from_axis))
      return l
```

File: tests/utils/test_maxtext_utils_nnx.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+from flax import nnx
+import jax.numpy as jnp
+from jax.sharding import PartitionSpec as P
+from maxtext.utils import maxtext_utils_nnx  # pylint: disable=no-name-in-module
+
+
+class TestMaxtextUtilsNnx(unittest.TestCase):
+  """Tests for maxtext_utils_nnx."""
+
+  def test_nnx_sync_moveaxis(self):
```
Comment (khatwanimohit) [inline tests/utils/test_maxtext_utils_nnx.py:27]: To ensure we prevent regressions for the standard MaxText pattern of omitting trailing `None`s, please add test cases with:
    1. `sharding = P("layers", "embed")` for `nnx_sync_moveaxis`.

File: tests/utils/test_maxtext_utils_nnx.py [hunk 7]
```diff
@@ -0,0 +1,72 @@
...
+    var.set_metadata(sharding=sharding, out_sharding=sharding)
+
+    moved_var = maxtext_utils_nnx.nnx_sync_moveaxis(var, 0, 2)
+
+    self.assertEqual(moved_var.get_value().shape, (20, 30, 10))
+    self.assertEqual(moved_var.get_metadata().get("sharding"), P("embed", None, "layers"))
+    self.assertEqual(moved_var.get_metadata().get("out_sharding"), P("embed", None, "layers"))
+
+  def test_nnx_remove_scan_axis(self):
+    val = jnp.zeros((10, 20))
```
Comment (khatwanimohit) [inline tests/utils/test_maxtext_utils_nnx.py:52]: 2. `sharding = P("embed")` for `nnx_add_scan_axis` (checking that it outputs `P("layers", "embed")` or equivalent)

File: src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt (hunk 4 above)
Comment (RexBearIU) [reply src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt]: Ideally yes, for maximum reproducibility. They are slightly mismatched here (`jax==0.10.2` vs `jaxlib ==0.10.1`) due to how dependencies are resolved during  `seed-env `:

- jaxlib is strictly pinned to a nightly commit (`a0d62932...`) via `seed-env`.

- jax is is resolved during the `seed-env`  run as an unpinned transitive dependency. Because `0.10.2` was just released, uv automatically resolved it to the newest version during the build.

File: src/maxtext/utils/maxtext_utils_nnx.py (hunk 5 above)
Comment (RexBearIU) [reply src/maxtext/utils/maxtext_utils_nnx.py:232]: Good catch. I've updated `move_fn` to pad the sharding list with trailing `None`s up to `val.ndim` before moving axes. This handles configurations where trailing `None`s are omitted.

File: tests/utils/test_maxtext_utils_nnx.py (hunk 6 above)
Comment (RexBearIU) [reply tests/utils/test_maxtext_utils_nnx.py:27]: Added `test_nnx_sync_moveaxis_missing_trailing` to verify. It sets up a 3D tensor with 2D `P("layers", "embed")` sharding and asserts the output becomes `P("embed", None, "layers")`.

File: tests/utils/test_maxtext_utils_nnx.py (hunk 7 above)
Comment (RexBearIU) [reply tests/utils/test_maxtext_utils_nnx.py:52]: Added `test_nnx_add_scan_axis_missing_trailing` to verify. It stacks a 2D tensor with 1D `P("embed")` sharding to 3D, and asserts that adding the scan axis pads and inserts `"layers"` to produce `P("layers", "embed", None)`. All 5 tests are passing.

**Review by khatwanimohit — APPROVED** [review]

File: src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt (hunk 4 above)
Comment (igorts-git) [reply src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt]: Unfortunately, we are currently seeing test failures with `jax==0.10.2` in our [nightly docker builds](https://github.com/AI-Hypercomputer/maxtext/actions/runs/28073431359/job/83113319563). I sent a message to the JAX team, but I would recommend against upgrading to this version.

Comment (SurbhiJainUSC) [reply src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt]: @igorts-git - why do you think our nightly build failure is related to 0.10.2?

Comment (RexBearIU) [reply src/dependencies/requirements/generated_requirements/tpu-post-train-requirements.txt]: Thanks for raising this. I tried updating `qwix` in both `tpu_post_train_overrides.txt` and `requirements.txt`, but both approaches still result in `uv` automatically upgrading JAX to `0.10.2` during regeneration. 

If we want to stay on `0.10.1`, I can explicitly pin `jax==0.10.1` and `jaxlib==0.10.1` in `tpu_post_train_overrides.txt` and regenerate. @SurbhiJainUSC @igorts-git let me know if you would like me to push those pins.
